internal/internal_task_handlers.go

```diff
@@ -1819,6 +1855,26 @@ func (ath *activityTaskHandlerImpl) Execute(taskList string, t *s.PollForActivit
 
 	ctx, span := createOpenTracingActivitySpan(ctx, ath.tracer, time.Now(), activityType, t.WorkflowExecution.GetWorkflowId(), t.WorkflowExecution.GetRunId())
 	defer span.Finish()
+
+	if activityImplementation.GetOptions().EnableAutoHeartbeat && t.HeartbeatTimeoutSeconds != nil && *t.HeartbeatTimeoutSeconds > 0 {
+		go func() {
+			autoHbInterval := time.Duration(*t.HeartbeatTimeoutSeconds) * time.Second / 2
+			ticker := time.NewTicker(autoHbInterval)
+			defer ticker.Stop()
+			for {
+				select {
+				case <-ctx.Done():
+					return
+				case <-ticker.C:
+					hbErr := invoker.BackgroundHeartbeat()
+					if IsCanceledError(hbErr) {
+						return
+					}
```

Comment:
My understanding is activity can ignore the cancellation and continue execution, in which case we should continue the background heartbeat?

Reply:
Ah, ok.  I didn't really understand that about cancellation but I think that makes sense if that's the case. If the cancelled activity keeps running but the heartbeat stops then a retry of the activity could be started while the original activity is still running?  Sounds like this should just ignore any errors and always keep running until the context is done?

Reply:
Change the behavior to continue heartbeat in case of error, until ctx is done.